### PR TITLE
PCC13 realnitinworks

### PR DESCRIPTION
**Pull request by realnitinworks**
Difficulty level (1-10): [6]
Estimated time spent (hours): [3]
Completed (yes/no): [Yes]
I stretched my coding skills (if yes what did you learn?): []
Other feedback (what can we improve?): []

### CODE REVIEW

File: 13/realnitinworks/directors.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                rating=movie['imdb_score']
+            )
+
+
+def from_year(movies, year=FROM_YEAR):
+    """ Filter movies based on year"""
+    for movie in movies:
+        if not year:
+            continue
+        if movie.year >= year:
```
Comment (bbelderbos) [inline 13/realnitinworks/directors.py]: could combine? `if year and movie.year >= year`

File: 13/realnitinworks/directors.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+def director_filter(movies, min_movies=MIN_MOVIES):
+    """ Filter movies whose their director's min movies """
+    director_count = Counter(
+        movie.director
+        for movie in movies
+    )
+
+    movies = from_year(convert_csv_to_dict())
+    return (
+        movie
```
Comment (bbelderbos) [inline 13/realnitinworks/directors.py]: I like how you layout your code, manual or `black` ?

File: 13/realnitinworks/directors.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    return director_movie_map
+
+
+def top_n_directors(director_movie_map, n=TOP_N):
+    """ Yield top n directors and their movie details """
+    def mean_rating(item):
+        _, details = item
+        return details.get("mean_rating")
+
+    sorted_directors = sorted(director_movie_map.items(), key=mean_rating, reverse=True)
```
Comment (bbelderbos) [inline 13/realnitinworks/directors.py]: probably manual, `black` would use multilines here?

File: 13/realnitinworks/directors.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    for idx, (director, details) in enumerate(sorted_directors, start=1):
```
Comment (bbelderbos) [inline 13/realnitinworks/directors.py]: `start=1` is neat

File: 13/realnitinworks/directors.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        len(movie.title)
+        for _, details in top_n_directors(directors, n=20)
+        for movie in details["movies"]
+    )
+
+    # Display the results
+    for idx, (director, details) in enumerate(top_n_directors(directors, n=20), start=1):
+        mean_rating = details["mean_rating"]
+        idx = idx < 10 and f"0{idx}" or f"{idx}"
+        print(f'{idx}. {director:<{max_movie_name_len+1}} {mean_rating}')
```
Comment (bbelderbos) [inline 13/realnitinworks/directors.py]: how much space f-strings save!

File: 13/realnitinworks/directors.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+            imdb_scores.append(float(movie.rating))
+        director_movie_map[director]["mean_rating"] = round(statistics.mean(imdb_scores), 1)
+    
+    return director_movie_map
+
+
+def top_n_directors(director_movie_map, n=TOP_N):
+    """ Yield top n directors and their movie details """
+    def mean_rating(item):
+        _, details = item
```
Comment (bbelderbos) [inline 13/realnitinworks/directors.py]: do we need intermediate variable? `item[1].get("mean_rating")`

File: 13/realnitinworks/directors.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    return directors
+
+
+def calculate_mean_imdb_score(director_movie_map):
+    """ Add mean_rating key into the input map"""
+    for director, value in director_movie_map.items():
+        imdb_scores = []
+        for movie in value["movies"]:
+            imdb_scores.append(float(movie.rating))
+        director_movie_map[director]["mean_rating"] = round(statistics.mean(imdb_scores), 1)
```
Comment (bbelderbos) [inline 13/realnitinworks/directors.py]: could keep it smaller by saving `round...` into a variable and assign the variable

File: 13/realnitinworks/directors.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        if director_count[movie.director] >= min_movies
+    )
+
+
+def map_director_movie(movies):
+    """ Make a mapping of directors and their movies"""
+    def by_rating(movie):
+        return movie.rating
+
+    directors = defaultdict(lambda: defaultdict(list))
```
Comment (bbelderbos) [inline 13/realnitinworks/directors.py]: `lambda` in`defaultdict ` is pretty wild

**Comment by bbelderbos** [comment]
great job

File: 13/realnitinworks/directors.py (hunk 1 above)
Comment (realnitinworks) [reply 13/realnitinworks/directors.py]: Oh! Why do I miss these 

> could combine? `if year and movie.year >= year`

Ohh! Why do I miss these . Thanks Bob

File: 13/realnitinworks/directors.py (hunk 6 above)
Comment (realnitinworks) [reply 13/realnitinworks/directors.py]: > do we need intermediate variable? `item[1].get("mean_rating")`

Agree. That is not required.

**Pull request merged**